### DISCUSSION TITLE
OKS Single instance

### DISCUSSION THREAD

**Discussion by Powerwoody** [talmolab/sleap#2180] [Help!]
Hello,

To me it seems a bit wired the way that mAP and mAR are calculated due to the fact that when we have a single animal instance per frame we dont need to match the keypoints to the correct instance in a frame, even if a entire set of prediction on a frame is nan then the oks would just 0 for that frame if we assume that gt points are annotated. Can someone clarify how oks is calculated in sleap more specifcally mAP and mAR and the recall curves and what the files: labels_pr.val.slp and labels_gt.val.slp actually contain after training. 

Thanks a lot.

Mvh. Marcus

**Comment by gitttt-1234** [comment]
Hi @Powerwoody !

Apologies for the delay.

- `labels_gt.val.slp`: ground truth labels for the validation set.

- `labels_pr.val.slp`: model predictions on that same set.

#### OKS (Object-keypoint similarity) 
OKS is computed by measuring the Euclidean distance between each predicted keypoint and its corresponding ground-truth keypoint. This distance is then normalized based on the scale of the object (bounding box area for the instance) and standard-deviation that defines the spread in the localization accuracy of each node. For each node, keypoint similarity is computed by taking the negative exponent of the normalized distance. Mean OKS is the average of keypoint similarities across all visible nodes.

Ref: [Ronch & Perona. "Benchmarking and Error Diagnosis in Multi-Instance Pose Estimation." ICCV (2017)](https://arxiv.org/abs/1707.05388).

#### VOC metrics (mAP, mAR)
The following VOC-style metrics are generated using either OKS or PCK as the matching scores and a set of thresholds where a predicted instance is considered as a True Postive if it's match score is greater than the threshold else it is counted as a False Positive.

- Average Precision (AP): Average of best precisions over fixed set of recall thresholds, at each match score threshold.
- Average Recall (AR): Maximum recall achieved at each match score threshold.
- Mean Average Precision (mAP): Mean of average precisions across match thresholds.
- Mean Average Recall (mAR): Mean of average recalls across match thresholds.

Hope this helps! Let me know if you have any questions!

Thanks, 

Divya
